packages/grid/_modules_/grid/utils/createSelector.ts

```diff
@@ -19,29 +21,48 @@ interface CreateSelectorFunction {
   ): OutputSelector<Result>;
 }
 
-const cache = {};
+const cache: Record<number | string, Map<any[], any>> = {};
 
 function isApiRef(stateOrApiRef: any): stateOrApiRef is GridApiRef {
   return stateOrApiRef.current;
 }
 
+let warnedOnce = false;
+
 export const createSelector: CreateSelectorFunction = (...args: any) => {
-  const selector = (stateOrApiRef: GridApiRef | GridState) => {
-    const cacheKey = isApiRef(stateOrApiRef) ? stateOrApiRef.current.instanceId : 'default';
+  const selector = (...selectorArgs: any[]) => {
+    const [stateOrApiRef, instanceId] = selectorArgs;
+    const cacheKey = isApiRef(stateOrApiRef)
+      ? stateOrApiRef.current.instanceId
+      : instanceId ?? 'default';
     const state = isApiRef(stateOrApiRef) ? stateOrApiRef.current.state : stateOrApiRef;
 
-    if (cache[cacheKey] && cache[cacheKey][args]) {
-      return cache[cacheKey][args](state);
+    if (process.env.NODE_ENV !== 'production') {
+      if (!warnedOnce && cacheKey === 'default') {
+        console.warn(
+          [
+            'MUI: A selector was called without passing the instance ID, which may impact the performance of the grid.',
+            'To fix, call it with `apiRef`, e.g. `mySelector(apiRef)`, or pass the instance ID explicitly, e.g `mySelector(state, apiRef.current.instanceId)`.',
+          ].join('\n'),
+        );
+        warnedOnce = true;
+      }
+    }
+
+    if (cache[cacheKey] && cache[cacheKey].get(args)) {
+      // We pass the cache key because the called selector might have as
+      // dependency another selector created with this `createSelector`.
+      return cache[cacheKey].get(args)(state, cacheKey);
     }
 
     const newSelector = reselectCreateSelector(...args);
 
     if (!cache[cacheKey]) {
-      cache[cacheKey] = {};
+      cache[cacheKey] = new Map();
```

Comment:
Do you know why the Map is more efficient here ?

Reply:
I think it's because the type of the keys. With a primitive as key (e.g number) the Map loses but wins when a complex value is used: https://jsbench.me/y4kz2z8xeb/1